src/Sylius/Bundle/ApiBundle/Validator/Constraints/ProductOrVariantEnabledValidator.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use Sylius\Bundle\ApiBundle\Command\Cart\AddItemToCart;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Product\Repository\ProductVariantRepositoryInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class ProductOrVariantEnabledValidator extends ConstraintValidator
+{
+    /** @var ProductVariantRepositoryInterface */
+    private $productVariantRepository;
+
+    public function __construct(ProductVariantRepositoryInterface $productVariantRepository)
+    {
+        $this->productVariantRepository = $productVariantRepository;
+    }
+
+    public function validate($value, Constraint $constraint)
+    {
+        Assert::isInstanceOf($value, AddItemToCart::class);
+
+        /** @var OrderProductEligibility $constraint */
+        Assert::isInstanceOf($constraint, ProductOrVariantEnabled::class);
+
+        /** @var ProductVariantInterface $productVariant */
+        $productVariant = $this->productVariantRepository->findOneBy(['code' =>$value->productVariantCode]);
+
+        if (!$productVariant->isEnabled()) {
+            $this->context->addViolation(
+                $constraint->message,
+                ['%productName%' => $productVariant->getName()]
+            );
+
+            return;
```

Comment:
Just to ask, would we like to have this early return here? Or we want to also validate if product is enabled?

Reply:
IMO early return is fine but in a different order - firstly you should check if the product is enabled and next check if a variant is enabled. Because if product is disabled you can't buy any variant regardless of his availability.